### PR TITLE
add Logger to expose event logging

### PR DESCRIPTION
**Pull request by ajwerner**
Circuit breakers currently expose very few mechanisms for observability.
There is an untested channel-based mechanism that could be used for a similar
purpose but this approach seemed simpler and more explicit. The hope is that
this additional information will be useful for debugging
github.com/cockroachdb/cockroach/issues/31361.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/circuitbreaker/3)

### CODE REVIEW

File: circuitbreaker.go [hunk 1]
```diff
@@ -190,10 +208,11 @@ func NewRateBreaker(rate float64, minSamples int64) *Breaker {
 
 // Subscribe returns a channel of BreakerEvents. Whenever the breaker changes state,
 // the state will be sent over the channel. See BreakerEvent for the types of events.
+// Note that events may be dropped or not sent so clients should not rely on
```
Comment (a-robinson) [inline circuitbreaker.go]: I think these comments accidentally got copied over from your other PR (ditto on the `AddListener` comment).

File: circuitbreaker.go [hunk 2]
```diff
@@ -336,7 +357,9 @@ func (cb *Breaker) Call(circuit func() error, timeout time.Duration) error {
 
 // CallContext is same as Call but if the ctx is canceled after the circuit returned an error,
 // the error will not be marked as a failure because the call was canceled intentionally.
-func (cb *Breaker) CallContext(ctx context.Context, circuit func() error, timeout time.Duration) error {
+func (cb *Breaker) CallContext(
```
Comment (a-robinson) [inline circuitbreaker.go]: If you're able to convince your editor not to do this, I'd rather not making meaningless formatting changes that make rebasing on any upstream changes harder.

File: circuitbreaker_test.go [hunk 3]
```diff
@@ -518,3 +518,30 @@ func TestPartialSecondBackoff(t *testing.T) {
 		t.Fatalf("expected breaker to be ready after more than nextBackoff time had passed")
 	}
 }
+
+func TestLogger(t *testing.T) {
```
Comment (a-robinson) [inline circuitbreaker_test.go:591]: Is this test actually testing anything? I'd expect it to verify that it received the expected event(s).

Comment (ajwerner) [reply circuitbreaker_test.go:591]: Oh man that was sloppy, I started then went and did something else and forgot I hadn't finished. Fixed now. Sorry for that.

File: circuitbreaker.go (hunk 2 above)
Comment (ajwerner) [reply circuitbreaker.go]: Done.

File: circuitbreaker.go (hunk 1 above)
Comment (ajwerner) [reply circuitbreaker.go]: Done.

**Comment by ajwerner** [comment]
@a-robinson TFTR, it's updated. Sorry for the sloppiness.

**Review by a-robinson — APPROVED** [review]
LGTM

**Comment by ajwerner** [comment]
TFTR!

**Pull request merged**